File: src/szu_netlogin/srun_crypto.py
"""Independent SRun BX1 field construction from the executable specification."""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import struct
from dataclasses import dataclass


STANDARD_BASE64_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
SRUN_BASE64_ALPHABET = "LVoJPiCN2R8G90yg+hmFHuacZ1OWMnrsSTXkYpUq/3dlbfKwv6xztjI7DeBE45QA"
_BASE64_TRANSLATION = str.maketrans(STANDARD_BASE64_ALPHABET, SRUN_BASE64_ALPHABET)


@dataclass(frozen=True, repr=False)
class SRunDerivedFields:
    password: str
    info: str
    checksum: str

    def __repr__(self) -> str:
        return "SRunDerivedFields(<redacted>)"


def hmac_md5_hex(password: str, challenge: str) -> str:
    return hmac.new(challenge.encode(), password.encode(), hashlib.md5).hexdigest()


def xencode(message: str, key: str) -> bytes:
    values = _to_uint32(message.encode("utf-8"), include_length=True)
    if not values:
        return b""
    key_words = _to_uint32(key.encode("utf-8"), include_length=False)
    key_words.extend([0] * (4 - len(key_words)))
    key_words = key_words[:4]
    count = len(values) - 1
    z = values[count]
    total = 0
    rounds = 6 + 52 // (count + 1)
    delta = 0x9E3779B9
    for _ in range(rounds):
        total = (total + delta) & 0xFFFFFFFF
        e = (total >> 2) & 3
        for index in range(count):
            y = values[index + 1]
            mixed = _mix(z, y, total, key_words[(index & 3) ^ e])
            values[index] = (values[index] + mixed) & 0xFFFFFFFF
            z = values[index]
        y = values[0]
        mixed = _mix(z, y, total, key_words[(count & 3) ^ e])
        values[count] = (values[count] + mixed) & 0xFFFFFFFF
        z = values[count]
    return b"".join(struct.pack("<I", value) for value in values)


def srun_base64(data: bytes) -> str:
    return base64.b64encode(data).decode("ascii").translate(_BASE64_TRANSLATION)


def stable_info_json(username: str, password: str, ip: str, acid: str) -> str:
    return json.dumps(
        {
            "username": username,
            "password": password,
            "ip": ip,
            "acid": acid,
            "enc_ver": "srun_bx1",
        },
        ensure_ascii=False,
        separators=(",", ":"),
    )


def derive_login_fields(
    *, username: str, password: str, ip: str, acid: str, challenge: str, n: str = "200", type_: str = "1"
) -> SRunDerivedFields:
    digest = hmac_md5_hex(password, challenge)
    info_json = stable_info_json(username, password, ip, acid)
    info = "{SRBX1}" + srun_base64(xencode(info_json, challenge))
    checksum_source = "".join(
        (
            challenge, username,
            challenge, digest,
            challenge, acid,
            challenge, ip,
            challenge, n,
            challenge, type_,
            challenge, info,
        )
    )
    return SRunDerivedFields(
        password="{MD5}" + digest,
        info=info,
        checksum=hashlib.sha1(checksum_source.encode()).hexdigest(),
    )


def _to_uint32(data: bytes, *, include_length: bool) -> list[int]:
    padding = (-len(data)) % 4
    words = list(struct.unpack(f"<{(len(data) + padding) // 4}I", data + b"\0" * padding)) if data else []
    if include_length:
        words.append(len(data))
    return words


def _mix(z: int, y: int, total: int, key: int) -> int:
    return (
        ((z >> 5) ^ ((y << 2) & 0xFFFFFFFF))
        + (((y >> 3) ^ ((z << 4) & 0xFFFFFFFF)) ^ (total ^ y))
        + (key ^ z)
    ) & 0xFFFFFFFF

File: macos/Sources/SZUNetCore/SRunCrypto.swift
import CryptoKit
import Foundation

public struct SRunDerivedFields: Equatable, Sendable {
    public let infoJSON: String
    public let hmacMD5Hex: String
    public let passwordField: String
    public let infoField: String
    public let checksum: String
}

public enum SRunCrypto {
    public static let base64Alphabet = "LVoJPiCN2R8G90yg+hmFHuacZ1OWMnrsSTXkYpUq/3dlbfKwv6xztjI7DeBE45QA"

    public static func deriveLoginFields(
        username: String,
        password: String,
        clientIP: String,
        acid: String,
        challenge: String,
        n: String = "200",
        type: String = "1"
    ) -> SRunDerivedFields {
        let hmac = hmacMD5Hex(password: password, challenge: challenge)
        let infoJSON = stableInfoJSON(
            username: username,
            password: password,
            clientIP: clientIP,
            acid: acid
        )
        let info = "{SRBX1}" + customBase64(xencode(message: infoJSON, key: challenge))
        let checksumInput = challenge + username
            + challenge + hmac
            + challenge + acid
            + challenge + clientIP
            + challenge + n
            + challenge + type
            + challenge + info
        return SRunDerivedFields(
            infoJSON: infoJSON,
            hmacMD5Hex: hmac,
            passwordField: "{MD5}" + hmac,
            infoField: info,
            checksum: sha1Hex(checksumInput)
        )
    }

    public static func stableInfoJSON(
        username: String,
        password: String,
        clientIP: String,
        acid: String
    ) -> String {
        "{\"username\":\(quoted(username)),\"password\":\(quoted(password)),"
            + "\"ip\":\(quoted(clientIP)),\"acid\":\(quoted(acid)),"
            + "\"enc_ver\":\"srun_bx1\"}"
    }

    public static func hmacMD5Hex(password: String, challenge: String) -> String {
        let key = SymmetricKey(data: Data(challenge.utf8))
        let code = HMAC<Insecure.MD5>.authenticationCode(
            for: Data(password.utf8),
            using: key
        )
        return hex(Data(code))
    }

    public static func sha1Hex(_ value: String) -> String {
        hex(Data(Insecure.SHA1.hash(data: Data(value.utf8))))
    }

    public static func xencode(message: String, key: String) -> Data {
        guard !message.isEmpty else { return Data() }
        var values = words(Data(message.utf8), includeLength: true)
        var keyWords = words(Data(key.utf8), includeLength: false)
        while keyWords.count < 4 { keyWords.append(0) }
        let last = values.count - 1
        var z = values[last]
        var sum: UInt32 = 0
        let delta: UInt32 = 0x9E37_79B9
        var rounds = 6 + 52 / values.count
        while rounds > 0 {
            sum &+= delta
            let e = Int((sum >> 2) & 3)
            for index in 0 ..< last {
                let y = values[index + 1]
                let mixed = mix(
                    z: z,
                    y: y,
                    sum: sum,
                    key: keyWords[(index & 3) ^ e]
                )
                values[index] &+= mixed
                z = values[index]
            }
            let y = values[0]
            let mixed = mix(
                z: z,
                y: y,
                sum: sum,
                key: keyWords[(last & 3) ^ e]
            )
            values[last] &+= mixed
            z = values[last]
            rounds -= 1
        }
        // The appended plaintext length word participates in encryption and is
        // ciphertext afterwards. Serialize every encrypted word; interpreting
        // the tail as a length truncates valid BX1 output.
        return bytes(values, includesLength: false)
    }

    private static func mix(z: UInt32, y: UInt32, sum: UInt32, key: UInt32) -> UInt32 {
        let first = (z >> 5) ^ (y << 2)
        let second = ((y >> 3) ^ (z << 4)) ^ (sum ^ y)
        return first &+ second &+ (key ^ z)
    }

    public static func customBase64(_ data: Data, alphabet: String = base64Alphabet) -> String {
        let table = Array(alphabet)
        precondition(table.count == 64)
        let bytes = Array(data)
        var output = ""
        var index = 0
        while index < bytes.count {
            let first = Int(bytes[index])
            let second = index + 1 < bytes.count ? Int(bytes[index + 1]) : 0
            let third = index + 2 < bytes.count ? Int(bytes[index + 2]) : 0
            output.append(table[first >> 2])
            output.append(table[((first & 3) << 4) | (second >> 4)])
            output.append(index + 1 < bytes.count ? table[((second & 15) << 2) | (third >> 6)] : "=")
            output.append(index + 2 < bytes.count ? table[third & 63] : "=")
            index += 3
        }
        return output
    }

    private static func quoted(_ value: String) -> String {
        let data = try! JSONEncoder().encode(value)
        return String(decoding: data, as: UTF8.self)
    }

    private static func words(_ data: Data, includeLength: Bool) -> [UInt32] {
        let bytes = Array(data)
        var result = [UInt32](repeating: 0, count: (bytes.count + 3) / 4)
        for (index, byte) in bytes.enumerated() {
            result[index >> 2] |= UInt32(byte) << UInt32((index & 3) * 8)
        }
        if includeLength { result.append(UInt32(bytes.count)) }
        return result
    }

    private static func bytes(_ words: [UInt32], includesLength: Bool) -> Data {
        guard !words.isEmpty else { return Data() }
        let payloadWords = includesLength ? words.dropLast() : words[...]
        var result = Data()
        result.reserveCapacity(payloadWords.count * 4)
        for word in payloadWords {
            result.append(UInt8(truncatingIfNeeded: word))
            result.append(UInt8(truncatingIfNeeded: word >> 8))
            result.append(UInt8(truncatingIfNeeded: word >> 16))
            result.append(UInt8(truncatingIfNeeded: word >> 24))
        }
        if includesLength, let requested = words.last {
            return Data(result.prefix(Int(requested)))
        }
        return result
    }

    private static func hex(_ data: Data) -> String {
        data.map { String(format: "%02x", $0) }.joined()
    }
}
